File: examples/primitives/uint/gas-calculations.zig
//! EVM Gas Calculations Example
//!
//! Demonstrates:
//! - Intrinsic gas costs
//! - Memory expansion costs
//! - EIP-1559 base fee calculations
//! - Gas optimization patterns

const std = @import("std");
const primitives = @import("primitives");
const Uint = primitives.Uint;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n=== EVM Gas Calculations Example ===\n\n", .{});

    // Gas constants
    const TX_BASE_COST: u256 = 21_000;
    const TX_DATA_ZERO_COST: u256 = 4;
    const TX_DATA_NONZERO_COST: u256 = 16;
    const MEMORY_COST_PER_WORD: u256 = 3;
    const QUADRATIC_DENOMINATOR: u256 = 512;
    const BASE_FEE_MAX_CHANGE: u256 = 8;
    const ELASTICITY_MULTIPLIER: u256 = 2;

    // 1. Intrinsic gas calculation
    std.debug.print("1. Intrinsic Gas Calculation\n", .{});
    std.debug.print("   ------------------------\n", .{});

    // Empty transaction
    const empty_gas = TX_BASE_COST;
    std.debug.print("   Empty transaction: {} gas\n\n", .{empty_gas});

    // Data with zeros
    const zero_count: u256 = 100;
    const zero_gas = TX_BASE_COST + (TX_DATA_ZERO_COST * zero_count);
    std.debug.print("   100 zero bytes: {} gas\n", .{zero_gas});
    std.debug.print("   ({} base + {} * 100)\n\n", .{ TX_BASE_COST, TX_DATA_ZERO_COST });

    // Data with non-zeros
    const nonzero_count: u256 = 100;
    const nonzero_gas = TX_BASE_COST + (TX_DATA_NONZERO_COST * nonzero_count);
    std.debug.print("   100 non-zero bytes: {} gas\n", .{nonzero_gas});
    std.debug.print("   ({} base + {} * 100)\n\n", .{ TX_BASE_COST, TX_DATA_NONZERO_COST });

    // ERC20 transfer (4 byte selector + mostly zeros)
    const erc20_nonzero: u256 = 4;
    const erc20_zero: u256 = 64; // Approximate
    const erc20_gas = TX_BASE_COST + (TX_DATA_NONZERO_COST * erc20_nonzero) + (TX_DATA_ZERO_COST * erc20_zero);
    std.debug.print("   ERC20 transfer: {} gas\n\n", .{erc20_gas});

    // 2. Memory expansion costs
    std.debug.print("2. Memory Expansion Costs\n", .{});
    std.debug.print("   ---------------------\n", .{});

    const expansions = [_]struct { from: u256, to: u256, desc: []const u8 }{
        .{ .from = 0, .to = 32, .desc = "0 → 32 bytes (1 word)" },
        .{ .from = 32, .to = 64, .desc = "32 → 64 bytes (2 words)" },
        .{ .from = 0, .to = 1024, .desc = "0 → 1024 bytes (32 words)" },
        .{ .from = 1024, .to = 10000, .desc = "1024 → 10000 bytes (significant)" },
    };

    for (expansions) |exp| {
        // Calculate memory cost: (words * 3) + (words^2 / 512)
        const new_words = (exp.to + 31) / 32;
        const old_words = (exp.from + 31) / 32;

        const new_linear = new_words * MEMORY_COST_PER_WORD;
        const old_linear = old_words * MEMORY_COST_PER_WORD;

        const new_quad = (new_words * new_words) / QUADRATIC_DENOMINATOR;
        const old_quad = (old_words * old_words) / QUADRATIC_DENOMINATOR;

        const new_total = new_linear + new_quad;
        const old_total = old_linear + old_quad;

        const cost = new_total - old_total;

        std.debug.print("   {s}: {} gas\n", .{ exp.desc, cost });
    }
    std.debug.print("\n", .{});

    // 3. EIP-1559 base fee calculation
    std.debug.print("3. EIP-1559 Base Fee Calculation\n", .{});
    std.debug.print("   ----------------------------\n", .{});

    const base_fee: u256 = 1_000_000_000; // 1 gwei
    const gas_limit: u256 = 30_000_000;
    const gas_target: u256 = gas_limit / ELASTICITY_MULTIPLIER;

    std.debug.print("   Current base fee: {} wei (1 gwei)\n", .{base_fee});
    std.debug.print("   Gas limit: {}\n", .{gas_limit});
    std.debug.print("   Gas target: {}\n\n", .{gas_target});

    const scenarios = [_]struct { used: u256, desc: []const u8 }{
        .{ .used = gas_target, .desc = "At target (no change)" },
        .{ .used = 29_000_000, .desc = "High usage (~97%)" },
        .{ .used = 1_000_000, .desc = "Low usage (~3%)" },
    };

    for (scenarios) |scenario| {
        const next_fee = calculateNextBaseFee(base_fee, scenario.used, gas_target);
        const percent_used = (@as(f64, @floatFromInt(scenario.used)) / @as(f64, @floatFromInt(gas_limit))) * 100.0;

        const change = if (next_fee > base_fee) next_fee - base_fee else base_fee - next_fee;
        const change_percent = (@as(f64, @floatFromInt(change)) / @as(f64, @floatFromInt(base_fee))) * 100.0;

        std.debug.print("   {s}:\n", .{scenario.desc});
        std.debug.print("   - Gas used: {} ({d:.1}%)\n", .{ scenario.used, percent_used });
        std.debug.print("   - Next base fee: {} wei\n", .{next_fee});
        std.debug.print("   - Change: {s}{d:.2}%\n\n", .{ if (next_fee > base_fee) "+" else "-", change_percent });
    }

    // 4. Gas optimization analysis
    std.debug.print("4. Gas Optimization Analysis\n", .{});
    std.debug.print("   ------------------------\n", .{});

    const SLOAD_COST: u256 = 2100;
    const SSTORE_SET: u256 = 20_000;
    const SSTORE_RESET: u256 = 5000;

    std.debug.print("   Storage operation costs:\n", .{});
    std.debug.print("   - SLOAD (cold): {} gas\n", .{SLOAD_COST});
    std.debug.print("   - SSTORE (set): {} gas\n", .{SSTORE_SET});
    std.debug.print("   - SSTORE (reset): {} gas\n\n", .{SSTORE_RESET});

    const num_reads: u256 = 10;
    const storage_reads = SLOAD_COST * num_reads;
    const memory_reads = 3 * num_reads;

    std.debug.print("   {} reads comparison:\n", .{num_reads});
    std.debug.print("   - From storage: {} gas\n", .{storage_reads});
    std.debug.print("   - From memory: {} gas\n", .{memory_reads});
    std.debug.print("   - Savings: {} gas\n\n", .{storage_reads - memory_reads});

    // 5. Complete transaction cost
    std.debug.print("5. Complete Transaction Cost\n", .{});
    std.debug.print("   -----------------------\n", .{});

    const tx_gas_price: u256 = 50_000_000_000; // 50 gwei
    const tx_gas_used: u256 = 100_000;
    const tx_value: u256 = 1_000_000_000_000_000_000; // 1 ETH

    const gas_cost = tx_gas_used * tx_gas_price;
    const total_cost = gas_cost + tx_value;

    const gas_cost_ether = @as(f64, @floatFromInt(gas_cost)) / 1e18;
    const total_cost_ether = @as(f64, @floatFromInt(total_cost)) / 1e18;

    std.debug.print("   Gas price: 50 gwei\n", .{});
    std.debug.print("   Gas used: {}\n", .{tx_gas_used});
    std.debug.print("   Transaction value: 1.000000 ETH\n\n", .{});

    std.debug.print("   Gas cost: {d:.6} ETH\n", .{gas_cost_ether});
    std.debug.print("   Total cost: {d:.6} ETH\n\n", .{total_cost_ether});

    // 6. Batch transaction savings
    std.debug.print("6. Batch Transaction Savings\n", .{});
    std.debug.print("   ------------------------\n", .{});

    const num_transfers: u256 = 10;
    const separate_cost = TX_BASE_COST * num_transfers;
    const batch_cost = TX_BASE_COST + (25_000 * (num_transfers - 1));

    std.debug.print("   {} separate transfers:\n", .{num_transfers});
    std.debug.print("   - Cost: {} gas\n\n", .{separate_cost});

    std.debug.print("   {} batched transfers:\n", .{num_transfers});
    std.debug.print("   - Cost: {} gas\n", .{batch_cost});
    std.debug.print("   - Savings: {} gas\n", .{separate_cost - batch_cost});
    std.debug.print("   - Percentage: {}%\n\n", .{((separate_cost - batch_cost) * 100) / separate_cost});

    std.debug.print("=== Example Complete ===\n\n", .{});
}

fn calculateNextBaseFee(current: u256, gas_used: u256, gas_target: u256) u256 {
    if (gas_used == gas_target) return current;

    if (gas_used > gas_target) {
        // Increase base fee
        const delta = gas_used - gas_target;
        const increase = (current * delta / gas_target) / 8;
        return @max(current + increase, 1);
    } else {
        // Decrease base fee
        const delta = gas_target - gas_used;
        const decrease = (current * delta / gas_target) / 8;
        return current - decrease;
    }
}

File: examples/primitives/uint/gas-calculations.ts
/**
 * Example: EVM Gas Calculations
 *
 * Demonstrates:
 * - Intrinsic gas costs
 * - Memory expansion costs
 * - EIP-1559 base fee calculations
 * - Gas optimization patterns
 */

import * as Uint from '../../../src/primitives/Uint/index.js';

console.log('\n=== EVM Gas Calculations Example ===\n');

// Gas constants
const TX_BASE_COST = Uint.from(21000n); // Base transaction cost
const TX_DATA_ZERO_COST = Uint.from(4n); // Per zero byte in calldata
const TX_DATA_NONZERO_COST = Uint.from(16n); // Per non-zero byte in calldata
const MEMORY_COST_PER_WORD = Uint.from(3n); // Linear memory cost
const QUADRATIC_DENOMINATOR = Uint.from(512n); // For memory expansion
const BASE_FEE_MAX_CHANGE = Uint.from(8n); // EIP-1559 max base fee change
const ELASTICITY_MULTIPLIER = Uint.from(2n); // EIP-1559 elasticity

// 1. Intrinsic gas calculation
console.log('1. Intrinsic Gas Calculation');
console.log('   ------------------------');

function calculateIntrinsicGas(data: Uint8Array): typeof Uint.prototype {
	let gas = TX_BASE_COST;

	for (const byte of data) {
		if (byte === 0) {
			gas = gas.plus(TX_DATA_ZERO_COST);
		} else {
			gas = gas.plus(TX_DATA_NONZERO_COST);
		}
	}

	return gas;
}

// Empty transaction
const emptyData = new Uint8Array(0);
const emptyGas = calculateIntrinsicGas(emptyData);
console.log(`   Empty transaction: ${emptyGas.toString()} gas\n`);

// Data with zeros
const zeroData = new Uint8Array(100); // All zeros
const zeroGas = calculateIntrinsicGas(zeroData);
console.log(`   100 zero bytes: ${zeroGas.toString()} gas`);
console.log(`   (${TX_BASE_COST.toString()} base + ${TX_DATA_ZERO_COST.toString()} * 100)\n`);

// Data with non-zeros
const nonZeroData = new Uint8Array(100).fill(1);
const nonZeroGas = calculateIntrinsicGas(nonZeroData);
console.log(`   100 non-zero bytes: ${nonZeroGas.toString()} gas`);
console.log(`   (${TX_BASE_COST.toString()} base + ${TX_DATA_NONZERO_COST.toString()} * 100)\n`);

// Mixed data (simulating ERC20 transfer calldata)
const erc20Transfer = new Uint8Array(68); // 4 byte selector + 32 bytes address + 32 bytes amount
erc20Transfer[0] = 0xa9; // transfer(address,uint256) selector
erc20Transfer[1] = 0x05;
erc20Transfer[2] = 0x9c;
erc20Transfer[3] = 0xbb;
// Rest are mostly zeros (typical for addresses and amounts)

const erc20Gas = calculateIntrinsicGas(erc20Transfer);
console.log(`   ERC20 transfer: ${erc20Gas.toString()} gas\n`);

// 2. Memory expansion costs
console.log('2. Memory Expansion Costs');
console.log('   ---------------------');

function calculateMemoryCost(newSize: typeof Uint.prototype, oldSize: typeof Uint.prototype): typeof Uint.prototype {
	// Memory cost = linear + quadratic
	// cost(size) = (size / 32) * 3 + (size / 32)^2 / 512

	const newWords = newSize.plus(Uint.from(31n)).dividedBy(Uint.from(32n));
	const oldWords = oldSize.plus(Uint.from(31n)).dividedBy(Uint.from(32n));

	// Linear cost
	const newLinear = newWords.times(MEMORY_COST_PER_WORD);
	const oldLinear = oldWords.times(MEMORY_COST_PER_WORD);

	// Quadratic cost
	const newQuad = newWords.times(newWords).dividedBy(QUADRATIC_DENOMINATOR);
	const oldQuad = oldWords.times(oldWords).dividedBy(QUADRATIC_DENOMINATOR);

	const newTotal = newLinear.plus(newQuad);
	const oldTotal = oldLinear.plus(oldQuad);

	// Return expansion cost (difference)
	return newTotal.minus(oldTotal);
}

const expansions = [
	{ from: 0n, to: 32n, desc: '0 → 32 bytes (1 word)' },
	{ from: 32n, to: 64n, desc: '32 → 64 bytes (2 words)' },
	{ from: 0n, to: 1024n, desc: '0 → 1024 bytes (32 words)' },
	{ from: 1024n, to: 10000n, desc: '1024 → 10000 bytes (significant)' },
];

for (const { from, to, desc } of expansions) {
	const cost = calculateMemoryCost(Uint.from(to), Uint.from(from));
	console.log(`   ${desc}: ${cost.toString()} gas`);
}
console.log();

// 3. EIP-1559 base fee calculation
console.log('3. EIP-1559 Base Fee Calculation');
console.log('   ----------------------------');

function calculateNextBaseFee(
	currentBaseFee: typeof Uint.prototype,
	gasUsed: typeof Uint.prototype,
	gasTarget: typeof Uint.prototype,
): typeof Uint.prototype {
	if (gasUsed.equals(gasTarget)) {
		return currentBaseFee;
	}

	if (gasUsed.greaterThan(gasTarget)) {
		// Block is above target - increase base fee
		const delta = gasUsed.minus(gasTarget);
		const increase = currentBaseFee.times(delta).dividedBy(gasTarget).dividedBy(BASE_FEE_MAX_CHANGE);
		return currentBaseFee.plus(increase).maximum(Uint.ONE);
	} else {
		// Block is below target - decrease base fee
		const delta = gasTarget.minus(gasUsed);
		const decrease = currentBaseFee.times(delta).dividedBy(gasTarget).dividedBy(BASE_FEE_MAX_CHANGE);
		return currentBaseFee.minus(decrease);
	}
}

const baseFee = Uint.from(1000000000n); // 1 gwei
const gasLimit = Uint.from(30000000n); // 30M gas limit
const gasTarget = gasLimit.dividedBy(ELASTICITY_MULTIPLIER); // 15M target

console.log(`   Current base fee: ${baseFee.toString()} wei (1 gwei)`);
console.log(`   Gas limit: ${gasLimit.toString()}`);
console.log(`   Gas target: ${gasTarget.toString()}\n`);

const scenarios = [
	{ used: gasTarget, desc: 'At target (no change)' },
	{ used: Uint.from(29000000n), desc: 'High usage (~97%)' },
	{ used: Uint.from(1000000n), desc: 'Low usage (~3%)' },
];

for (const { used, desc } of scenarios) {
	const nextFee = calculateNextBaseFee(baseFee, used, gasTarget);
	const percentUsed = (Number(used.toBigInt()) / Number(gasLimit.toBigInt())) * 100;
	const change = nextFee.greaterThan(baseFee) ? nextFee.minus(baseFee) : baseFee.minus(nextFee);
	const changePercent = Number(change.times(Uint.from(10000n)).dividedBy(baseFee).toBigInt()) / 100;

	console.log(`   ${desc}:`);
	console.log(`   - Gas used: ${used.toString()} (${percentUsed.toFixed(1)}%)`);
	console.log(`   - Next base fee: ${nextFee.toString()} wei`);
	console.log(`   - Change: ${nextFee.greaterThan(baseFee) ? '+' : '-'}${changePercent.toFixed(2)}%\n`);
}

// 4. Gas optimization patterns
console.log('4. Gas Optimization Analysis');
console.log('   ------------------------');

// Storage operations
const SLOAD_COST = Uint.from(2100n); // Cold SLOAD
const SSTORE_SET = Uint.from(20000n); // Set storage from zero
const SSTORE_RESET = Uint.from(5000n); // Reset storage to zero

console.log(`   Storage operation costs:`);
console.log(`   - SLOAD (cold): ${SLOAD_COST.toString()} gas`);
console.log(`   - SSTORE (set): ${SSTORE_SET.toString()} gas`);
console.log(`   - SSTORE (reset): ${SSTORE_RESET.toString()} gas\n`);

// Compare storage vs memory
const numReads = Uint.from(10n);
const storageReads = SLOAD_COST.times(numReads);
const memoryReads = Uint.from(3n).times(numReads); // ~3 gas per memory read

console.log(`   ${numReads.toString()} reads comparison:`);
console.log(`   - From storage: ${storageReads.toString()} gas`);
console.log(`   - From memory: ${memoryReads.toString()} gas`);
console.log(`   - Savings: ${storageReads.minus(memoryReads).toString()} gas\n`);

// 5. Transaction cost estimation
console.log('5. Complete Transaction Cost');
console.log('   -----------------------');

const txGasPrice = Uint.from(50000000000n); // 50 gwei
const txGasUsed = Uint.from(100000n);
const txValue = Uint.from(1000000000000000000n); // 1 ETH

const gasCost = txGasUsed.times(txGasPrice);
const totalCost = gasCost.plus(txValue);

function weiToEther(wei: typeof Uint.prototype): string {
	return (Number(wei.toBigInt()) / 1e18).toFixed(6);
}

console.log(`   Gas price: 50 gwei`);
console.log(`   Gas used: ${txGasUsed.toString()}`);
console.log(`   Transaction value: ${weiToEther(txValue)} ETH\n`);

console.log(`   Gas cost: ${weiToEther(gasCost)} ETH`);
console.log(`   Total cost: ${weiToEther(totalCost)} ETH\n`);

// 6. Batch transaction savings
console.log('6. Batch Transaction Savings');
console.log('   ------------------------');

const singleTxCost = TX_BASE_COST;
const numTransfers = Uint.from(10n);
const separateCost = singleTxCost.times(numTransfers);

// Batch transfer saves base cost for each additional transfer
const batchCost = TX_BASE_COST.plus(Uint.from(25000n).times(numTransfers.minus(Uint.ONE)));

console.log(`   ${numTransfers.toString()} separate transfers:`);
console.log(`   - Cost: ${separateCost.toString()} gas\n`);

console.log(`   ${numTransfers.toString()} batched transfers:`);
console.log(`   - Cost: ${batchCost.toString()} gas`);
console.log(`   - Savings: ${separateCost.minus(batchCost).toString()} gas`);
console.log(`   - Percentage: ${Number(separateCost.minus(batchCost).times(Uint.from(100n)).dividedBy(separateCost).toBigInt())}%\n`);

console.log('=== Example Complete ===\n');
